### PR TITLE
Auto-refresh UserInfoPopup recent messages

### PR DESCRIPTION
**Pull request by dnsge**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description
Automatically add new messages to UserInfoPopup dialogs as they are received without having to press the refresh button. Clicking refresh will load the latest messages again from the channel.

<details>
  <summary>Video examples</summary>

[Example 1](https://i.nuuls.com/z3w_-.mp4)
[Example 2](https://i.nuuls.com/o517H.mp4)
</details>

### CODE REVIEW

File: src/widgets/dialogs/UserInfoPopup.cpp [hunk 1]
```diff
@@ -53,6 +53,23 @@ namespace {
 
         return label.getElement();
     };
+
+    bool checkMessage(const QString &userName, MessagePtr message)
```
Comment (leon-richardt) [inline src/widgets/dialogs/UserInfoPopup.cpp:57]: nit: can this have a more descriptive name?

File: src/widgets/dialogs/UserInfoPopup.cpp [hunk 2]
```diff
@@ -53,6 +53,23 @@ namespace {
...
+    {
+        bool isSubscription = message->flags.has(MessageFlag::Subscription) &&
+                              message->loginName == "" &&
```
Comment (23rd) [inline src/widgets/dialogs/UserInfoPopup.cpp:60]:
```suggestion
                              message->loginName.isEmpty() &&
```

File: src/widgets/dialogs/UserInfoPopup.cpp [hunk 3]
```diff
@@ -53,6 +53,23 @@ namespace {
 
         return label.getElement();
     };
+
+    bool checkMessage(const QString &userName, MessagePtr message)
+    {
```
Comment (23rd) [inline src/widgets/dialogs/UserInfoPopup.cpp:58]:
```suggestion
    {
        if (message->flags.has(MessageFlag::Whisper))
            return false;

```

File: src/widgets/dialogs/UserInfoPopup.cpp [hunk 4]
```diff
@@ -53,6 +53,23 @@ namespace {
...
+                              message->messageText.split(" ").at(0).compare(
+                                  userName, Qt::CaseInsensitive) == 0;
+
+        bool isModAction =
+            message->timeoutUser.compare(userName, Qt::CaseInsensitive) == 0;
+        bool isSelectedUser =
+            message->loginName.compare(userName, Qt::CaseInsensitive) == 0;
+
+        return (isSubscription || isModAction || isSelectedUser) &&
+               !message->flags.has(MessageFlag::Whisper);
```
Comment (23rd) [inline src/widgets/dialogs/UserInfoPopup.cpp:70]:
```suggestion
        return (isSubscription || isModAction || isSelectedUser);
```

File: src/widgets/dialogs/UserInfoPopup.cpp [hunk 5]
```diff
@@ -462,6 +466,29 @@ void UserInfoPopup::updateLatestMessages()
...
+    else
+    {
+        this->refreshConnection_ =
+            this->channel_->messageAppended.connect([this](auto message, auto) {
+                if (checkMessage(this->userName_, message))
+                {
+                    this->updateLatestMessages();
+                }
+            });
+    }
```
Comment (23rd) [inline src/widgets/dialogs/UserInfoPopup.cpp:491]:
```suggestion
    this->refreshConnection_ = this->channel_->messageAppended.connect(
        [this, hasMessages](auto message, auto) {
            if (!checkMessage(this->userName_, message))
                return;
            if (hasMessages)
                this->ui_.latestMessages->channel()->addMessage(message);
            else
                this->updateLatestMessages();
        });
```

File: src/widgets/dialogs/UserInfoPopup.cpp (hunk 1 above)
Comment (dnsge) [reply src/widgets/dialogs/UserInfoPopup.cpp]: Fixed in 91c3889 :+1:

File: src/widgets/dialogs/UserInfoPopup.cpp (hunk 2 above)
Comment (dnsge) [reply src/widgets/dialogs/UserInfoPopup.cpp]: This and your other suggestions are fixed in 91c3889 :+1:

**Comment by Mm2PL** [comment]
Messages in usercards are duplicated, probably because I have the same channel open in multiple splits. Please fix that :slightly_smiling_face:

**Comment by dnsge** [comment]
I couldn't reproduce the duplicated messages with having 2 of the same channel open. Could you upload a video of the issue?

**Review by fourtf — COMMENTED** [review]
Signal doesn't get disconnected when the pop-up closes? Or am I seeing wrong

**Comment by fourtf** [comment]
Looks good to me 👍 Thank you for your contribution!

**Comment by dnsge** [comment]
@fourtf I was doing some testing and found that `this->setAttribute(Qt::WA_DeleteOnClose);` needs to be added in the constructor otherwise the window won't be deleted if closed by the `x` button.

**Comment by fourtf** [comment]
@dnsge I'll push a fix in a sec

**Pull request merged**